drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,352 @@
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, config->desc->enable_val);
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to enable regulator");
+	}
+
+	return ret;
+}
+
+static int axp192_disable(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	int ret = 0;
+
+	LOG_INST_DBG(config->log, "Disabling regulator");
+	LOG_INST_DBG(config->log, "[0x%02x]=0 mask=0x%x", (uint32_t)config->desc->enable_reg,
+		     (uint32_t)config->desc->enable_mask);
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, 0u);
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to disable regulator");
+	}
+
+	return ret;
+}
+
+static unsigned int axp192_count_voltages(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return linear_range_group_values_count(config->desc->ranges, config->desc->num_ranges);
+}
+
+static int axp192_list_voltage(const struct device *dev, unsigned int idx, int32_t *volt_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges, idx,
+					    volt_uv);
+}
+
+static int axp192_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	uint16_t idx = 0;
+	int ret = 0;
+
+	LOG_INST_DBG(config->log, "voltage = [min=%d, max=%d]", min_uv, max_uv);
+
+	/* set voltage */
+	ret = linear_range_group_get_win_index(config->desc->ranges, config->desc->num_ranges,
+					       min_uv, max_uv, &idx);
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "No voltage range window could be deteceted");
+		return ret;
+	}
+
+	idx <<= config->desc->vsel_bitpos;
+
+	LOG_INST_DBG(config->log, "[0x%x]=0x%x mask=0x%x", (uint32_t)config->desc->vsel_reg, idx,
+		     (uint32_t)config->desc->vsel_mask);
+	ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->vsel_reg, config->desc->vsel_mask,
+				     (uint8_t)idx);
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to set regulator voltage");
+	}
+
+	return ret;
+}
+
+static int axp192_get_voltage(const struct device *dev, int32_t *volt_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	int ret;
+	uint8_t raw_reg;
+
+	/* read voltage */
+	ret = i2c_reg_read_byte_dt(&config->i2c, config->desc->vsel_reg, &raw_reg);
+	if (ret != 0) {
+		return ret;
+	}
+
+	raw_reg = (raw_reg & config->desc->vsel_mask) >> config->desc->vsel_bitpos;
+
+	ret = linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges, raw_reg,
+					   volt_uv);
+
+	return ret;
+}
+
+static int axp192_set_mode(const struct device *dev, regulator_mode_t mode)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	int ret = 0;
+
+	/* setting workmode is only possible for DCDC1-3 */
+	if (config->desc->workmode_reg != 0) {
```

Comment:
this will return 0 for devices not supporting it, so maybe
```suggestion
	/* setting workmode is only possible for DCDC1-3 */
	if (config->desc->workmode_reg != 0) {
	    return -ENOTSUP;
	 }
```

Reply:
omg, what a stupid bug! :-(

Reply:
fixed in [4a22568](https://github.com/zephyrproject-rtos/zephyr/pull/58809/commits/4a22568b18d1602f41a2e5ae50c38f6de57add25)